cypress/integration/x-msc2716-historical-import/historical-import.spec.ts

```diff
@@ -0,0 +1,628 @@
+        const userIdWithoutServer = virtualUserId.split(':')[0];
+        const localpart = userIdWithoutServer.replace(/^@/, '');
+        return localpart;
+    });
+
+    virtualUserLocalparts.forEach((virtualUserLocalpart) => {
+        cy.request<{ error?: string, errcode?: string }>({
+            url,
+            method: "POST",
+            body: {
+                type: "m.login.application_service",
+                username: virtualUserLocalpart,
+            },
+            headers: {
+                'Authorization': `Bearer ${applicationServiceToken}`,
+            },
+            // We'll handle the errors ourselves below
+            failOnStatusCode: false,
+        })
+            .then((res) => {
+                // Registration success
+                if (res.status === 200) {
+                    return;
+                }
+
+                const errcode = res.body.errcode;
+
+                // User already registered and good to go
+                if (res.status == 400 && errcode === "M_USER_IN_USE") {
+                    return;
+                }
+
+                const errorMessage = res.body.error;
+                throw new Error(`ensureVirtualUserRegistered failed to register: (${errcode}) ${errorMessage}`);
+            });
+    });
+}
+
+function sendMarkerEventAndEnsureHistoryDetectedStatusBar(asMatrixClient) {
+    // Send the marker event which lets the client know there are
+    // some historical messages back at the given insertion event.
+    cy.all([
+        cy.get<string>("@roomId"),
+        cy.get<string>("@baseInsertionEventId"),
+    ]).then(async ([roomId, baseInsertionEventId]) => {
+        const { event_id: markeEventId } = await asMatrixClient.sendStateEvent(
+            roomId,
+            'org.matrix.msc2716.marker', {
+                "org.matrix.msc2716.marker.insertion": baseInsertionEventId,
+            },
+            Cypress._.uniqueId("marker_state_key_"),
+        );
+
+        cy.wrap(markeEventId).as('markeEventId');
+
+        // Wait for the message to show up for the logged in user
+        waitForEventIdsInClient([markeEventId]);
+    });
+
+    // Ensure the "History import detected" notice is shown
+    cy.get(`[data-cy="historical-import-detected-status-bar"]`).should("exist");
+}
+
+function setupRoomWithHistoricalMessagesAndMarker({
+    synapse,
+    asMatrixClient,
+    virtualUserIDs,
+}: {
+    synapse: SynapseInstance;
+    asMatrixClient: MatrixClient;
+    virtualUserIDs: string[];
+}) {
+    // As the application service, create the room so it is the room creator
+    // and proper power_levels to send MSC2716 events. Then join the logged
+    // in user to the room.
+    cy.wrap(null).then(async function() {
+        const resp = await asMatrixClient.createRoom({
+            // FIXME: I can't use Preset.PublicChat because Cypress doesn't
+            // understand Typescript to import it
+            preset: "public_chat" as Preset,
```

Comment:
What should I do here?

Reply:
It should be used from `window.matrixcs` like the other enum usages

Reply:
👍 Thanks!